### ISSUE TITLE
[BUG] Spamassassin seems not to work when using LDAP

### ISSUE THREAD

**Issue by williamdes** [docker-mailserver/docker-mailserver#2123]
# Bug Report

### What is affected by this bug?

All emails

### When does this occur?

Email received

### How do we replicate the issue?

?

## Behavior

Headers should be added

### Actual Behavior

No headers

### Expected Behavior

Headers

## Your Environment

- version: 10.1.0

### Environment Variables

https://github.com/desportes/infrastructure/blob/2a637784845bc7b51621671f36aec96b8d55abaa/docker-compose.yml#L45

See my user patch to enable the headers:

https://github.com/desportes/infrastructure/blob/2a637784845bc7b51621671f36aec96b8d55abaa/user-patches.sh#L34-L35

## TLDR

There is an override that prevents Spamassassin to work: `/etc/amavis/conf.d/60-dms_default_config`
Read: https://github.com/docker-mailserver/docker-mailserver/issues/2123#issuecomment-907890537

**Comment by williamdes** [comment]
Currently investigating this because I NEVER did see an email flagged as spam nor automatically moved to Spams.

In the logs I can see nowhere about spamassasin startup

This is strange, currently hacking the scripts to find out

NOTE: I do not understand the `_notify` helper script, how to turn it verbose ? (by reading the function I was clueless, so I added an echo to hack that)

**Comment by williamdes** [comment]
Current debug
```
mailserver_1    | task/Setting up Security Stack/
mailserver_1    | inf/Enabling and configuring spamassassin/
mailserver_1    | inf/Configure Spamassassin/Amavis to put SPAM inbox/
mailserver_1    | inf/Enabling clamav/
mailserver_1    | inf/Fail2ban enabled/
mailserver_1    | inf/Amavis enabled/
mailserver_1    | task/Setting up Postfix Aliases/
mailserver_1    | inf/Configuring root alias/
mailserver_1    | task/Setting up Postfix vhost/
mailserver_1    | task/Setting up Postfix dhparam/
mailserver_1    | inf/Use ffdhe4096 for dhparams (postfix)/
mailserver_1    | inf/Configuring postscreen/
mailserver_1    | inf/Configuring postfix message size limit/
mailserver_1    | inf/Configuring postfix mailbox size limit/
mailserver_1    | inf/Configuring postfix virtual mailbox size limit/
mailserver_1    | inf/Configuring user access/
mailserver_1    | task/Setting up Postfix Override configuration/
```

Seems like the config step is done
Amavis is logged on startup so I assume the service starts also
But for now still searching about why I find no traces of this amavis/spamassasin in my emails

**Comment by williamdes** [comment]
For LDAP setups I bet `/etc/amavis/conf.d/05-domain_id` is somewhat wrong

Also `/etc/amavis/conf.d/60-dms_default_config` that refers to `/etc/postfix/vhost`

My server is at xxx.servers.datacenters.network so the values are filled with `servers.datacenters.network` and that value is, well, useless and incorrect

**Comment by casperklein** [comment]
How did you confirm, that **no** spam at all is detected?

What happens, when you send yourself an email with the content ``XJS*C4JDBQADN1.NSBN3*2IDNEN*GTUBE-STANDARD-ANTI-UBE-TEST-EMAIL*C.34X``?

**Comment by williamdes** [comment]
I had broke the amavis config by forgetting to escape the `@` in my new config, now beck to normal it helped proove the emails pass into amavis. `status=deferred (connect to 127.0.0.1[127.0.0.1]:10024: Connection refused)`

I now have some errors like that
```
Aug 11 22:23:39 mail amavis[2339]: (02339-01) (!)connect to /var/run/clamav/clamd.ctl failed, attempt #1: Can't connect to a UNIX socket /var/run/clamav/clamd.ctl: No such file or directory
```

I recall having them from time to time but never did really pay attention

For the record I always had `X-Virus-Scanned: Yes` headers but nothing about spam.

> How did you confirm, that no spam at all is detected?

I receive spam all day long, and never not one email got to Junk. This is why today I tried to add the spam check headers but they are not adding in my received emails. So I can only conclude that there is no spam check

**Comment by williamdes** [comment]
Just added more config for spamassasin in https://github.com/desportes/infrastructure/commit/5cac7f618652bcf7aa7e8409ff4b9d6ace60d885

Using: http://blog.dmitryleskov.com/small-hacks/forcing-spamassassin-to-add-the-x-spam-status-header-to-ham-for-debugging/

But still no headers, continuing to dig to find where is the missing bolt

**Comment by williamdes** [comment]
2hours and 30 min later, impossible to get the spam headers to work on the mails I receive. And I did not find any evidence amavis cares about spamassasin.

It seems to ignore it totally, for an unknown reason. But it searches for viruses, I clearly do not care about, it's more to protect Windows users.

```
mailserver_1    | Aug 12 00:20:31 mail amavis[2329]: (02329-02) Passed CLEAN {RelayedOpenRelay}, [10.10.10.8]:39514 <williamdes@xxxxxxxxxx> -> <xxxxxxxxxxx@gmail.com>, Queue-ID: 0990042DAA, Message-ID: <a2069c68644d2f0d6278f90c85c67425@xxxxxxxxx>, mail_id: PEzNLS8leqCs, Hits: -0.201, size: 1743, queued_as: A690E42DBB, dkim_sd=mail:xxxxxxxxxx, 5535 ms
mailserver_1    | Aug 12 00:20:31 mail postfix/smtp[2822]: 0990042DAA: to=<xxxxxxxxxxxxxx@gmail.com>, relay=127.0.0.1[127.0.0.1]:10024, delay=5.7, delays=0.16/0.01/0.01/5.5, dsn=2.0.0, status=sent (250 2.0.0 from MTA(smtp:[127.0.0.1]:10025): 250 2.0.0 Ok: queued as A690E42DBB)
```

**Comment by williamdes** [comment]
Yesterday I enabled logwatch and pflogsumm
```
 --------------------- Amavisd-new Begin ------------------------ 

       20   *Warning: Virus scanner connection failure
       10   Miscellaneous warnings 
 
      143   Total messages scanned ------------------  100.00%
    3.874M  Total bytes scanned                      4,061,759
 ========   ==================================================
 
      143   Passed ----------------------------------  100.00%
        8     Spam passed                                5.59%
      135     Clean passed                              94.41%
 ========   ==================================================
 
        8   Spam ------------------------------------    5.59%
        8     Spam passed                                5.59%
 
      135   Ham -------------------------------------   94.41%
      135     Clean passed                              94.41%
 ========   ==================================================
 
 
 ---------------------- Amavisd-new End ------------------------- 
```

I can see it talks about spam but the rest is still a mystery

**Comment by georglauterbach** [comment]
You can turn notify verbose by enabling `DMS_DEBUG=1`:)

**Comment by williamdes** [comment]
I really need help, can someone please help me find out?

**Comment by georglauterbach** [comment]
Maybe @casperklein can help you further with the cryptic `XJSC4JDBQADN1.NSBN32IDNENGTUBE-STANDARD-ANTI-UBE-TEST-EMAILC.34X` message.

I'd personally try this:

1. Add more capabilities
2. Remove the `sed ...` line in `user-patches.sh` to see whether this may interfere with SpamAssasin
3. Maybe try a non-LDAP setup to prove your hypothesis about `/etc/amavis/conf.d/05-domain_id`

**Comment by casperklein** [comment]
Sorry, I don't have an idea at the moment. Regarding the "cryptic" message:

>The GTUBE ("Generic Test for Unsolicited Bulk Email") is a 68-byte test string used to test anti-spam systems, in particular those based on SpamAssassin. In SpamAssassin, it carries an antispam score of 1000 by default, which would be sufficient to trigger any installation.

https://en.wikipedia.org/wiki/GTUBE

**Comment by williamdes** [comment]
For now I did not have time to debug this
Maybe I will spend so of my week end to try to find out

But clearly, I asked some of my users and nobody reported seeing spam detection

It's not a big deal for them as they do not have their email in public databases.

But mine is all over the web so I get spam all day, and nothing gets detected.
That said virus detection seems to work because I did receive reports about received malware attachements

Will let you know, if anyone can run a self test by sending to self a mail with the magic subject mentioned earlier and say if it detects spam that would help narrow the issue

Also I am reading about open ports and am not sure if I should let 25 open
Does anyone know about if it is usefull nowadays? 
https://mailtrap.io/blog/smtp-ports-25-465-587-used-for/

**Comment by casperklein** [comment]
> .. and am not sure if I should let 25 open

Closing port 25 will ultimately block SPAM, but also all other legitimate mails 😉

**Comment by williamdes** [comment]
I used https://docker-mailserver.github.io/docker-mailserver/v10.0/examples/tutorials/basic-installation/

```yaml
version: '2'

services:
 mail:
    image: mailserver/docker-mailserver:latest
    hostname: mail
    domainname: example.org
    container_name: mail
    ports:
    - "25:25"
    - "587:587"
    - "465:465"
    volumes:
    - ./data/:/var/mail/
    - ./state/:/var/mail-state/
    - ./config/:/tmp/docker-mailserver/
    environment:
    - PERMIT_DOCKER=network
    #- SSL_TYPE=letsencrypt
    - ONE_DIR=1
    - DMS_DEBUG=1
    - SPOOF_PROTECTION=0
    - REPORT_RECIPIENT=1
    - ENABLE_SPAMASSASSIN=1
    - ENABLE_CLAMAV=1
    - ENABLE_FAIL2BAN=0
    - ENABLE_POSTGREY=0
    cap_add:
    - NET_ADMIN
    - SYS_PTRACE
```

```sh
docker-compose up -d
wget https://raw.githubusercontent.com/docker-mailserver/docker-mailserver/v10.1.2/setup.sh
chmod a+x ./setup.sh
./setup.sh email add admin@example.org passwd123
./setup.sh email add info@example.org passwd123
docker-compose restart
```

(you can use `telnet` instead of `nc` and seed the input by a copy paste)
```sh
echo 'EHLO localhost
MAIL FROM:<info@example.org>
RCPT TO:<admin@example.org>
DATA
MIME-Version: 1.0
Subject: XJSC4JDBQADN1.NSBN32IDNENGTUBE-STANDARD-ANTI-UBE-TEST-EMAILC.34X
From: <info@example.org>
To: <admin@example.org>
Date: Wed, 28 Apr 2021 11:42:10 +0000
Content-Type: text/html; charset="UTF-8"
Content-Transfer-Encoding: quoted-printable

To: <admin@example.org>

<b>Cool email body</b>

.
QUIT
' | nc localhost 25
220 mail.example.org ESMTP
250-mail.example.org
250-PIPELINING
250-SIZE 10240000
250-ETRN
250-STARTTLS
250-AUTH PLAIN LOGIN
250-AUTH=PLAIN LOGIN
250-ENHANCEDSTATUSCODES
250-8BITMIME
250-DSN
250 CHUNKING
250 2.1.0 Ok
250 2.1.5 Ok
354 End data with <CR><LF>.<CR><LF>
250 2.0.0 Ok: queued as F365BB00A9C
221 2.0.0 Bye
```

`./data/example.org/admin/new/*`

```
Return-Path: <info@example.org>
Delivered-To: admin@example.org
Received: from mail.example.org
        by mail.example.org with LMTP
        id rWE0DGiXK2F8GAAAwX/9vA
        (envelope-from <info@example.org>)
        for <admin@example.org>; Sun, 29 Aug 2021 14:19:20 +0000
Received: from localhost (localhost [127.0.0.1])
        by mail.example.org (Postfix) with ESMTP id 2502EB00A9C
        for <admin@example.org>; Sun, 29 Aug 2021 14:19:20 +0000 (UTC)
X-Virus-Scanned: Yes
X-Spam-Flag: NO
X-Spam-Score: 2.951
X-Spam-Level: **
X-Spam-Status: No, score=2.951 tagged_above=2 required=6.31
        tests=[ALL_TRUSTED=-1, DATE_IN_PAST_96_XX=2.07, HTML_MESSAGE=0.001,
        HTML_MIME_NO_HTML_TAG=0.635, MIME_HTML_ONLY=1.105, MISSING_MID=0.14]
        autolearn=no autolearn_force=no
Received: from localhost (unknown [172.28.0.1])
        by mail.example.org (Postfix) with ESMTP id 95FE9B00A9B
        for <admin@example.org>; Sun, 29 Aug 2021 14:19:18 +0000 (UTC)
MIME-Version: 1.0
Subject: XJSC4JDBQADN1.NSBN32IDNENGTUBE-STANDARD-ANTI-UBE-TEST-EMAILC.34X
From: <info@example.org>
To: <admin@example.org>
Date: Wed, 28 Apr 2021 11:42:10 +0000
Content-Type: text/html; charset="UTF-8"
Content-Transfer-Encoding: quoted-printable
Message-Id: <20210829141920.2502EB00A9C@mail.example.org>

To: <admin@example.org>

<b>Cool email body</b>
```

So, it works on this minimal local setup without LDAP..

Note: this doc example has a bad ident, and uses a deprecated ENV `REPORT_RECIPIENT`

**Comment by georglauterbach** [comment]
So this is an LDAP issue?

**Comment by williamdes** [comment]
> So this is an LDAP issue?

Probably, will try to dig into it
I would really love anyone having a LDAP setup trying the spam thing
That would really help me a lot

**Comment by georglauterbach** [comment]
I would like to help, but I'm not using LDAP.

**Comment by williamdes** [comment]
Before
```
 Passed CLEAN {RelayedOpenRelay}, [xx.xx.xx.xx]:43872 [xx.xx.xx.xx] <xxx@gmail.com> -> <xxx@datacenters.network>
 ```
 After removing `/etc/amavis/conf.d/60-dms_default_config`
 ```
 Passed CLEAN {RelayedInbound}, [xx.xx.xx.xx]:50504 [xx.xx.xx.xx] <xxx@gmail.com> -> <xxx@datacenters.network>
 ```
 
 This made me search on this path: https://groups.google.com/g/mailing.unix.amavis-user/c/QH9nxYQwjN4?pli=1
 
 The FAQ lied to me 😢 
 ```
 The default setup `@local_domains_acl = ( ".$mydomain" );` does not match subdomains.
 To match subdomains, you can override the `@local_domains_acl`
 directive in the amavis user config file `50-user` with `@local_domains_maps = (".");` to match any sort of domain template.
```


Since https://github.com/docker-mailserver/docker-mailserver/pull/563 the FAQ is wrong and the override can not work
Was to fix: https://github.com/docker-mailserver/docker-mailserver/issues/496

**Comment by georglauterbach** [comment]
That a big no-no, and we should fix that. Can you solve this @williamdes?

**Comment by williamdes** [comment]
> That a big no-no, and we should fix that. Can you solve this @williamdes?

Yes, but I m not sure what strategy to adopt

**Comment by georglauterbach** [comment]
> > That a big no-no, and we should fix that. Can you solve this @williamdes?
> 
> Yes, but I m not sure what strategy to adopt

I understand. My advise would be to use the strategy that you think fits best. In the end, you can submit a PR, state the reasons for the change and we well review it, so in theory, nothing should go wrong.

**Comment by williamdes** [comment]
Seems like I will run in a work overload for some time, anything will have delay
It's an estimation but I wanted to let you know

**Comment by casperklein** [comment]
Could you please update your initial post, with a short "tl;dr", that contains the new findings. So that not everyone have to read the whole thread 😉.

**Comment by matrixes** [comment]
I'm grateful for your meticulous look into this @williamdes because I am in the same situation; there are no added headers or any other indication that mail is being scanned and processed for spam; all spam blast straight through to my inboxes.

I'll have a look at my config and test the tldr `60-dms...`-modification suggested and report back with any findings.

**Comment by dumblob** [comment]
I wanted to try this container in LDAP setup but this issue scares me kind of. Is it still valid?

**Comment by georglauterbach** [comment]
We do not have too many maintainers using LDAP. Maybe you can try and report back here @dumblob?

**Comment by SnowMB** [comment]
I just ran into this issue when one of my users complained that he receives to much spam, while on my own account everything works fine. 

After testing extensively I can confirm that the default settings just work for the primary domain the mail-server is on. Additional domains are not "protected" by spamassassin. This has to do with the default setting `60-dms_default_config` being wrong.

```
use strict;

@local_domains_maps = (
   read_hash('/etc/postfix/vhost')
);

1;  # ensure a defined return
```

This can not now all domains I configured in ldap.

The [FAQ](https://docker-mailserver.github.io/docker-mailserver/edge/faq/#why-are-spamassassin-x-headers-not-inserted-into-my-subdomainexamplecom-subdomain-emails) mentions a workaround considering all domains as local by changing the `50-user` config: 

```
[Why are SpamAssassin x-headers not inserted into my subdomain.example.com subdomain emails?](https://docker-mailserver.github.io/docker-mailserver/edge/faq/#why-are-spamassassin-x-headers-not-inserted-into-my-subdomainexamplecom-subdomain-emails)

In the default setup, amavis only applies SpamAssassin x-headers into domains matching the template listed in the config file (05-domain_id in the amavis defaults).

The default setup @local_domains_acl = ( ".$mydomain" ); does not match subdomains. To match subdomains, you can override the @local_domains_acl directive in the amavis user config file 50-user with @local_domains_maps = ("."); to match any sort of domain template.
```

This gave me some trouble to figure out.... Of course this never takes any effect because the `60-dms_default_config` file is still present and will directly overwrite the setting in `50-user` :facepalm: .

However replacing `60-dms_default_config` with the workaround finally made it work :partying_face: 

# tl;dr

If you are using multiple domains via ldap configuration and want to have spam detection for all domains use 
`@local_domains_maps = ["."];` in `/etc/amavis/conf.d` but make sure to place it so, that config files with higher numbers don't overwrite it!

# extra tip
If you mount the config-file from your host system make sure the user and group are correct otherwise amavis will not start correctly!

**Comment by cloud-erik** [comment]
@SnowMB you are my hero of the day. It took me a whole day to get apamassin now finaly with LDAP up and running. Thanks for description.

**Comment by williamdes** [comment]
I am closing this one as I will not have time to continue tracking this discussion. My current patches seem to work over the years.
